Site Editor: Add ability to focus on editing a page's content vs the page's template

noisysocks opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Second attempt at https://github.com/WordPress/gutenberg/pull/49980. The key difference is that it's based on https://github.com/WordPress/gutenberg/pull/50643 instead of https://github.com/WordPress/gutenberg/pull/50082.

Requires https://github.com/WordPress/gutenberg/pull/50643.

Part of https://github.com/WordPress/gutenberg/issues/49404 and https://github.com/WordPress/gutenberg/issues/44461.

Adds functionality to the Site Editor that lets you, when editing a page, switch between focusing on editing the page's content versus editing the page's template.

## Why?
Adding some separation helps to prevent a very common point of frustration encountered when a user attempts to change something on a specific page but the change is made everywhere.

## How?
- Add breadcrumb and sidebar UI to `wp-edit-site` that lets the user toggle between focusing on editing the template and focusing on editing the page.
- Use the new `useBlockEditingMode()` hook added in https://github.com/WordPress/gutenberg/pull/50643 to prevent editing blocks that aren't the Post Title, Post Featured Image and Post Content block.
- Various UI affordances.
  - Notifications to nudge the user to edit the template when they click on a locked block.
  - A custom right sidebar for when the focus is on editing the page.

## Testing Instructions
1. Ensure that you have a primary Navigation menu saved.
1. Go to _Appearance → Editor → Navigation_ and select a page.
1. Test the new switching functionality.
1. Check that navigating to _Appearance → Editor → Templates_ and selecting a template results in the same site editor experience as before (no switching functionality).

## Screenshots or screencast 

https://user-images.githubusercontent.com/612155/233555590-9036ad7c-e3ad-47dd-8861-8058bb7168d4.mp4

## Follow up tasks

- [x] Rename all the methods I'm adding to refer to `post` instead of `page`. 
- [x] DRY up `ContentBlocksList` and `BlockInspectorLockedBlocks`.
- [ ] Explore opportunities for user education e.g. display a guide when switching to the template.
- [x] Make Post Title, Post Featured Image, Post Content blocks say "Page" instead of "Post" https://github.com/WordPress/gutenberg/issues/48963
- [ ] Explore making `useBlockEditingMode()` available for extenders. (Mark it public, add docs, add dev note.)
- [ ] A black pixel appears where the toolbar used to appear when Post Title is selected. 
- [ ] Try to remove `__unstable` `contentOnly` methods. (Not really related but using this list as my todo list now 😂)
- [ ] Make the notifications appear in the top centre with a purple background. As in the design mockup.
- [ ]  [Make it so that the global inserter inserts content into the Post Content block when no block is selected.](https://github.com/WordPress/gutenberg/issues/51538#issuecomment-1593097498)
- [x] Remove `select()` hack from `getBlockEditingMode` when https://github.com/WordPress/gutenberg/pull/50985 is fixed.
- [ ] Optimise selectors https://github.com/WordPress/gutenberg/pull/50643#pullrequestreview-1450920100 
- [ ] Add an E2E test
- [ ] https://github.com/WordPress/gutenberg/issues/51259
- [ ] Add 'Edit template' and 'Back to page' commands to the command centre

noisysocks committed:
Add ability to prevent editing blocks using useBlockEditingMode()

noisysocks committed:
Make useBlockEditingMode use context

noisysocks committed:
Remove rootBlockEditingMode setting

noisysocks committed:
Fix private createRegistrySelector selectors

noisysocks committed:
Consolidate templateLock=contentOnly logic into getBlockEditingMode

noisysocks committed:
Hide disabled blocks from List View

noisysocks committed:
Hide disabled blocks from breadcrumbs

noisysocks committed:
Add doc comments

noisysocks committed:
Add unit tests

noisysocks committed:
Use @typedef to document mode param

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into try/block-editing-mode

noisysocks committed:
Restore packages/components/package.json from trunk

noisysocks committed:
Restore packages/block-library/src/post-title/edit.js from trunk

noisysocks committed:
Move BlockListBlockContext out of block.js so that it exists on mobile platforms

noisysocks committed:
Site Editor: Add ability to focus on editing a page's content vs the page's template

noisysocks commented:
I just need to redo the breadcrumb component (up top) and content list component (in the sidebar) and then this should be ready.

fabiankaegy commented:
Thank you for working on this! As stated before, I think this can have a huge potential to not just impact the site editor but also allow for previewing the template in the post editor.

I took this PR in its current form for a spin and found have some things I noticed:

1. If the Post Featured Image block for example is nested inside the header for example it is only selectable via the sidebar. I think the issue here might be that unselected template parts have the overlay so that they clearly showcase that they are a template parts. But therefore that interferes with this new feature. (See video)

https://github.com/WordPress/gutenberg/assets/20684594/e7248b78-e69f-4740-9f76-ee349d45d0fd

2. The notice to edit the template only appears once at the very beginning. After that there seemingly is no way to get that message to re-appear and it is very hard to understand how one can actually get to the template that affects this page. 

Also even with jumping to the template it somehow still didn't quite register what template exactly I was currently editing and therefore which areas of my site would be affected by the change.

3. I do think 3rd party blocks should also be able to use this mechanism. As an example, if we think about a WooCommerce product price block that reads/writes post meta, that block should also be editable in this special mode.

talldan commented on packages/edit-site/src/components/sidebar-edit-mode/settings-header/index.js:
```diff
@@ -1,67 +1,90 @@
+				? // translators: ARIA label for the Template sidebar tab, selected.
+				  __( 'Page (selected)' )
+				: // translators: ARIA label for the Template Settings Sidebar tab, not selected.
+				  __( 'Page' );
```
It might be nice to replicate this label in the block breadcrumbs at the bottom, as they still say 'Template' when the sidebar says 'Page'.

noisysocks replied:
Good idea!

talldan commented:
Small observation as well, but really only partially related to what this PR achieves. It might seem weird to users that the blocks are called Post Title, Post Featured Image and Post Content when they're editing a Page. That's our implementation leaking into the user's experience. A lot of beginner users won't know that a page is a type of post.

noisysocks committed:
Show page information in DocumentActions

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into add/template-post-context-switching

noisysocks committed:
Implement Content panel

mikemcalister commented:
Great job on this, @noisysocks! This is providing some much needed visual separation between post and template editing. 

A few notes that may be helpful. When you click the Edit Template link in the snack bar notification, it instantly switches over to the edit template view. IMO, the switch over is almost too instant and not obvious enough. It's hard to tell you switched into a different view, even though the snack bar says you're in a new view. 

I wonder if a subtle animation between the views would help the user see this transition better and give them a much-needed visual cue that a view change has happened. 

Alternatively, maybe the first time this template edit view is triggered, we throw the small feature intro modal (similar to the Welcome Guide) to really grab their attention and be intentional about this transition. 

One thing I've learned over the past few years of working with users in the block editor is that subtlety is often lost! I err on the side of obvious and intentional when possible.

noisysocks committed:
Prevent block overlay on disabled blocks

noisysocks committed:
Fix Navigation block being selectable

noisysocks committed:
Show 'Page' in breadcrumbs when focused on editing page

noisysocks commented:
Thanks for the feedback everyone!

I added the missing pieces now and https://github.com/WordPress/gutenberg/pull/50643 is merged so this should be nearly ready to go.

> If the Post Featured Image block for example is nested inside the header for example it is only selectable via the sidebar. I think the issue here might be that unselected template parts have the overlay so that they clearly showcase that they are a template parts. But therefore that interferes with this new feature. (See video)

Good catch! I've fixed this in [4489da2](https://github.com/WordPress/gutenberg/pull/50857/commits/4489da2a9491a36f1b05bf3020e19c0c83fd7919). It doesn't make sense for a disabled block to ever have a block overlay since you can't select a disabled block.

> The notice to edit the template only appears once at the very beginning. After that there seemingly is no way to get that message to re-appear and it is very hard to understand how one can actually get to the template that affects this page.

We intentionally only show the notifications once, but maybe that's misguided. What do you think @SaxonF?

> Also even with jumping to the template it somehow still didn't quite register what template exactly I was currently editing and therefore which areas of my site would be affected by the change.

This should be a little bit better now that I'm showing the template name in the bar at the top of the editor in [727ecd2](https://github.com/WordPress/gutenberg/pull/50857/commits/727ecd2365150f3202ac5ddd3ae1c38b42e47d3f).

> I do think 3rd party blocks should also be able to use this mechanism. As an example, if we think about a WooCommerce product price block that reads/writes post meta, that block should also be editable in this special mode.

~Good idea. I'll make the array that we use filterable.~ **edit:** No I won't, see comment below.

> Small observation as well, but really only partially related to what this PR achieves. It might seem weird to users that the blocks are called Post Title, Post Featured Image and Post Content when they're editing a Page. That's our implementation leaking into the user's experience. A lot of beginner users won't know that a page is a type of post.

- [x] Good flag. I'll try to fix this. It's trickier than I first thought because you don't have access to block context in `__experimentalLabel`.

> A few notes that may be helpful. When you click the Edit Template link in the snack bar notification, it instantly switches over to the edit template view. IMO, the switch over is almost too instant and not obvious enough. It's hard to tell you switched into a different view, even though the snack bar says you're in a new view.
>
> I wonder if a subtle animation between the views would help the user see this transition better and give them a much-needed visual cue that a view change has happened.

Maybe an animation between states in the bar at the top of the editor would help with this? What do you think @SaxonF?

> Alternatively, maybe the first time this template edit view is triggered, we throw the small feature intro modal (similar to the Welcome Guide) to really grab their attention and be intentional about this transition.

I like the idea of adding more education. Not sure about using `WelcomeGuide` though since it's very all-encompassing. Let's explore this in a follow-up PR so that we can iterate on a few ideas.

noisysocks committed:
Update post title, post featured image, and post content blocks to say 'Page' instead of 'Post'

noisysocks commented:
> > I do think 3rd party blocks should also be able to use this mechanism. As an example, if we think about a WooCommerce product price block that reads/writes post meta, that block should also be editable in this special mode.
>
> Good idea. I'll make the array that we use filterable.

On second thought, a filter here isn't a good idea as filters aren't reactive. If the allowed block types returned by the filter changes, then the editor won't be updated to reflect that.

I think the better approach is for third party blocks to call `useBlockEditingMode( 'default' )` or `useBlockEditingMode( 'contentOnly' )` to override the inherited `'disabled'` state. They can do this either in a `BlockEdit` filter (which is what the site editor does) or by just updating the block's `edit` function. This hook is private at the moment but I can see us making it public after this feature gets some testing.

noisysocks commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/content-blocks-list.js:
```diff
@@ -0,0 +1,78 @@
+import { getBlockType, __experimentalGetBlockLabel } from '@wordpress/blocks';
+import { store as blockEditorStore, BlockIcon } from '@wordpress/block-editor';
+
+// TODO: This overlaps a lot with BlockInspectorLockedBlocks in
+// @wordpress/block-editor. DRY them into a single component.
```
Will do this in a follow-up. I don't think it belongs in this PR.

noisysocks commented:
Cool this is taking shape. I think I've addressed all of y'all feedback. Keen to get it in asap so that there's time for iteration before 6.3.

@SaxonF is going to push some tweaks to the top bar.

@ramonjd is going to review everything tomorrow.

Three little bugs I noticed just then that I'll fix tomorrow:

- [ ] ~A black pixel appears where the toolbar used to appear when Post Title is selected.~ **edit:** will do this in a follow-up
- [x] Double clicking on a locked block selects it.
- [x] Block appender displays at the bottom of the page when locked.

fabiankaegy commented:
> > > I do think 3rd party blocks should also be able to use this mechanism. As an example, if we think about a WooCommerce product price block that reads/writes post meta, that block should also be editable in this special mode.
> > 
> > 
> > Good idea. I'll make the array that we use filterable.
> 
> On second thought, a filter here isn't a good idea as filters aren't reactive. If the allowed block types returned by the filter changes, then the editor won't be updated to reflect that.
> 
> I think the better approach is for third party blocks to call `useBlockEditingMode( 'default' )` or `useBlockEditingMode( 'contentOnly' )` to override the inherited `'disabled'` state. They can do this either in a `BlockEdit` filter (which is what the site editor does) or by just updating the block's `edit` function. This hook is private at the moment but I can see us making it public after this feature gets some testing.

Yeah, I agree that the ideal solution would be for the `useBlockEditingMode` API to be made available publically :) I think it is great we can test it as a private feature in the Gutenberg plugin first. However, I think if this feature is to make it into WordPress core, that API would need to already be publically available.

It would be a real bummer to ship this feature in core without the ability for plugins to support this mode.

SaxonF committed:
toolbar title styling

noisysocks committed:
Remove removePostFromContentBlockLabels for now

noisysocks commented:
> > Small observation as well, but really only partially related to what this PR achieves. It might seem weird to users that the blocks are called Post Title, Post Featured Image and Post Content when they're editing a Page. That's our implementation leaking into the user's experience. A lot of beginner users won't know that a page is a type of post.
>
> Good flag. I'll try to fix this. It's trickier than I first thought because you don't have access to block context in __experimentalLabel.

I had fixed this in 8c1c4a6f221935750c2e98c496976be85808bff9 but decided to revert it in 23837deb7e97deab15e87a94fb8c79f3025c6730 because I just came across https://github.com/WordPress/gutenberg/issues/48963 and think it would be wiser to address this more holistically as a follow-up. For instance all custom post types should be supported not just Post and Page.

> It would be a real bummer to ship this feature in core without the ability for plugins to support this mode.

Totally agree. It's a difficult balance when introducing new APIs. We want to empower extenders but also once an API is in Core we can never really change it. I saw that @getdave is using this API for an unrelated purpose so maybe he can provide some feedback on the design. If it's good then we can stabilise it in time for 6.3. I've noted this as a follow-up.

noisysocks committed:
Fix being able to select text within disabled blocks

noisysocks committed:
Hide BlockAppender when block is disabled

ramonjd commented on packages/block-editor/src/components/block-list/content.scss:
```diff
@@ -178,12 +169,19 @@
+	pointer-events: auto;
+	user-select: text;
+
+	// But disable when editing is disabled.
```
Are these comments useful? Or are the CSS rules enough to explain?

noisysocks replied:
Yeah I suppose it's just noise 👍

ramonjd commented on packages/edit-site/src/components/block-editor/index.js:
```diff
@@ -137,6 +146,7 @@ export default function BlockEditor() {
 		contentRef,
 		useClipboardHandler(),
 		useTypingObserver(),
+		usePageContentLockNotifications(),
```
UX question for follow up:  The notification in the bottom right hand corner is confusing to me, mainly because it hangs around too long. When I edit a template, then select a fresh page without clicking on "Back to page", it's still there telling me I'm editing a template.

noisysocks replied:
True. In @SaxonF's mockup the notification appeared in the top centre of the page and was purple. I do like that design better. I'll add this to the follow-up list.

tellthemachines commented on packages/block-library/src/post-title/edit.js:
```diff
@@ -114,20 +119,22 @@ export default function PostTitleEdit( {
+						} }
+					/>
+				</BlockControls>
+			) }
```
I'm still able to edit/interact with block toolbars in the header when in content editing mode, though the toolbars no longer show the block type:

 
<img width="273" alt="Screenshot 2023-05-25 at 1 27 28 pm" src="https://github.com/WordPress/gutenberg/assets/8096000/a5d0a0c0-a3a9-4184-ba4e-3d454a462751">

<img width="275" alt="Screenshot 2023-05-25 at 1 27 57 pm" src="https://github.com/WordPress/gutenberg/assets/8096000/966b8e4a-e156-40d4-ad37-a49467fb764f">

noisysocks replied:
Huh, I don't see that... what steps did you take?

talldan replied:
I can too - site title and paragraphs in the header and footer seem to be editable. Not sure I did anything special.

I also noticed that when switching from editing the template back to editing the page content, the block you have selected might be part of the template, but it remains selected and has a toolbar.

andrewserong replied:
I noticed the same. This PR is otherwise testing quite nicely for me, though!

Here's a screengrab in case it helps:

https://github.com/WordPress/gutenberg/assets/14988353/d7e22204-454b-40d6-8d3a-9d42260936ba

noisysocks replied:
Do you have the latest changes pulled? There was a bug where you could select a locked block by double clicking or dragging that should be fixed in 3cd4a17.

andrewserong replied:
> Do you have the latest changes pulled?

Yes, just tried re-pulling, and my environment says it's up to date, and I re-ran `npm run dev`. In the screengrab, I'm single-clicking the Site Title block in the header, but I can also reproduce by clicking the Site Title or Paragraph blocks in the footer, too.

Very weird — I can see the `pointer-events: none` rule should be applying, but I was still able to select the block 🤔

<img width="1116" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/b7756d14-ac36-41c1-b369-868009274719">

talldan replied:
I have latest changes too.

noisysocks replied:
Are you using TT3?

andrewserong replied:
Yep!

noisysocks replied:
Aha, looks like it only happens in Chrome. I'll investigate tomorrow.

ramonjd commented on packages/edit-site/src/components/header-edit-mode/document-actions/index.js:
```diff
@@ -11,18 +16,68 @@ import {
+
+	const { togglePageContentLock } = useDispatch( editSiteStore );
+
+	// Return a simple loading indicator until we have information to show.
```
Is this comment still valid?

noisysocks replied:
I'm not sure. Pretty useless comments though. I'll remove.

tellthemachines commented on packages/edit-site/src/components/page-content-lock/use-disable-non-content-blocks.js:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Disables non-content blocks using the `useBlockEditingMode` hook.
+ */
+export function useDisableNonContentBlocks() {
```
Might the block-editor package be a better home for this? It would be super useful to have this template/content distinction for post editor template editing too (are there plans to do that at all?)

noisysocks replied:
Right now we only surface pages in the site editor via the W menu. I could see us eventually surfacing posts as well, even if it's only via navigating your site in browse mode on the right hand side. Even more long term I could see the site editor replacing the post editor.

This would all still be in the `edit-site` package, though, so having `useDisableNonContentBlocks` in `block-editor` wouldn't help in any way.

fabiankaegy replied:
From my perspective even today this api could be used to load a preview of the template in the post editor like described in #27847 

So even before an eventual unification I think there is great value in this starting now for a very large user base of every WordPress user that isn't an administrator. Because currently the site editor is limited to administrators.

tellthemachines replied:
> This would all still be in the edit-site package, though, so having useDisableNonContentBlocks in block-editor wouldn't help in any way.

I meant for the template editor that lives in edit-post, the one you can access through editing the template in the post editor sidebar. It would be nice if it worked like this too.

Might also be worth considering if this functionality might be useful for non-WP block editor instances.

noisysocks replied:
I feel that the `edit-site` editor is a much better starting place for building a unified editor (one that can switch between posts and templates seamlessly) than the `edit-post` editor. The `edit-site` editor is already able to handle multiple entity types and the `edit-post` editor contains a lot of legacy stuff e.g. custom fields. In other words I think a better alternative to https://github.com/WordPress/gutenberg/issues/27847 (and along the lines of https://github.com/WordPress/gutenberg/issues/41717) is to expand the functionality we're adding in this PR to the point where we can remove the post editor entirely in lieu of the site editor for block themes. As part of that we would have to make `site-editor.php` more intelligent about what capabilities it requires so that non-admins can use it.

But getting ahead of myself. This is all very medium-long term and not all that relevant to this PR which is starting out with a fairly limited feature.

It's far easier to move something from `edit-site` to `block-editor` in the future than the other way around due to backwards compatibility concerns, so I'm comfortable with `useDisableNonContentBlocks` starting out here.

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/content-blocks-list.js:
```diff
@@ -0,0 +1,78 @@
+				! [
+					'core/post-title',
+					'core/post-featured-image',
+					'core/post-content',
```
Not important, but maybe this array could be a const, exported from packages/edit-site/src/components/page-content-lock/use-disable-non-content-blocks.js so they're always synched?

noisysocks replied:
Aye yes I meant to do this. Thanks, done.

tellthemachines commented on packages/edit-site/src/components/sidebar-edit-mode/sidebar-card/style.scss:
```diff
@@ -0,0 +1,34 @@
+.edit-site-sidebar-card {
```
No strong feelings about it, but why aren't we using `Flex` or `HStack` and the like instead?

noisysocks replied:
Good point. I'm hesitant to change the implementation very much in this PR. I've tried to limit the changes to just moving the existing code as-is to a new reusable component.

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/content-blocks-list.js:
```diff
@@ -0,0 +1,78 @@
+					'core/post-content',
+				].includes( blockName )
+			) {
+				return [];
```
Is there any benefit to creating a stable reference here? E.g., `EMPTY_ARRAY` rather than recreating the array?

noisysocks replied:
Check my understanding but I don't think so because the callback passed to `useSelect` should only be called when the deps array (`[]`) changes?

ramonjd replied:
I just double-checked, looks like if deps is an empty array ([]) the callback is memoized.

`contentBlocks === contentBlocks` gives me `true` every time. 

🍺

tellthemachines reviewed (commented):
This is working pretty smoothly on testing, just a couple issues/questions below!

Also, I notice that the inserter no longer works inside the Navigation block in this branch.

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/index.js:
```diff
@@ -0,0 +1,93 @@
+				<ContentBlocksList />
+			</PanelBody>
+			<PanelBody
+				className="edit-site-edit-template-panel"
```
```suggestion
				className="edit-site-sidebar-edit-mode-page-panel__template"
```

No a biggie. It's a mouthful, but it follows established pattern. Take it or leave it.

noisysocks replied:
Ugh hate this stuff 😀

I'll change the selectors to `edit-site-page-panels__*` since `PagePanels` is the component and it's not really tied to `SidebarEditMode`.

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/index.js:
```diff
@@ -0,0 +1,93 @@
+			>
+				<VStack>
+					<div>{ getTemplateTitle() }</div>
+					<div className="edit-site-edit-template-panel__preview">
```
```suggestion
					<div className="edit-site-sidebar-edit-mode-page-panel__template-preview">
```

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/index.js:
```diff
@@ -0,0 +1,93 @@
+						</BlockContextProvider>
+					</div>
+					<Button
+						className="edit-site-edit-template-panel__button"
```
```suggestion
						className="edit-site-sidebar-edit-mode-page-panel__button"
```

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/index.js:
```diff
@@ -0,0 +1,93 @@
+					<Button
+						className="edit-site-edit-template-panel__button"
+						variant="secondary"
+						onClick={ () => togglePageContentLock() }
```
```suggestion
						onClick={ () => togglePageContentLock( false ) }
```

We can be explicit here about the [toggle value](https://github.com/WordPress/gutenberg/blob/599a4d1ab24d295371232cb992cc274043099612/packages/edit-site/src/store/actions.js#L542-L542) if we're just toggling things off.

ramonjd reviewed (approved):
✅  When editing a page in the site editor for the first time, only the content blocks (title, featured image and post content) are enabled
✅  Toggling "edit template" or using the notification link toggles `hasPageContentLock` and I can edit the entire template (but not the content blocks)
✅  Breadcrumb footer and sidebar reflect current editing state, e.g., page vs template blocks
✅  I can edit a page in the site editor and the contents are saved as I'd expect
✅  I checked template lock functionality again 

I left a few non-blocking comments.

ramonjd commented:
Performance test error could be related (looking for `.edit-site-canvas-spinner`). Kicked the tests off again to see.

noisysocks committed:
Fix comments block in non-post templates

noisysocks committed:
Update template card selector in E2E tests

noisysocks committed:
Fix 'Add submenu' button in Navigation block

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into add/template-post-context-switching

noisysocks committed:
Remove unhelpful comments

noisysocks committed:
Remove more unnecessary comments

noisysocks committed:
Use constant for block types array

noisysocks committed:
Use BEMish selectors

noisysocks committed:
Be explicit that we're switching away from the page lock

andrewserong commented:
This is generally testing really well for me!

One other issue I ran into is that it seems to be possible to re-arrange the Post Content, Post Title, and Post Featured Image blocks, and even drag Post Title into Post Content:

https://github.com/WordPress/gutenberg/assets/14988353/7336f0d9-5248-4a0f-9e25-a1d8fcd2d00d

Is it possible to treat these blocks in this context as if they've been set to have movement disabled? I.e. these settings:

<img width="469" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/01d05acd-a779-4932-a97a-0d1120679ba0">

Probably not a blocker to landing this PR, though, and could be looked at in a follow-up if need be?

noisysocks commented:
Getting close. Hopefully tests pass now. There's a few small things I'd like to do before merge:

- [x] Fix this problem @talldan found:

> I also noticed that when switching from editing the template back to editing the page content, the block you have selected might be part of the template, but it remains selected and has a toolbar.

- [ ] Rename all the methods I'm adding to refer to `post` instead of `page`. In WordPress we usually always call the underlying model a _post_ (even when it's a page). That should make things clearer if we ever expand this feature to cover posts and other CPTs. I also want to move away from using "lock" terminology since this doesn't have anything to do with `templateLock` anymore.

- [x] Disable the root inserter for now. In a follow-up I'll try to make it so that it is enabled but inserts content into the Post Content block.

- [x] Fix drag and drop in List View (see @andrewserong's comment above).

- [ ] Fix selection issue in Chrome. https://github.com/WordPress/gutenberg/pull/50857#discussion_r1205058523

noisysocks committed:
Update removePageFromBlockContext() test

noisysocks committed:
Fix post context from appearing in Edit Template preview

We have to set postId and postType to null for the preview and omit them
in the editor. No point having helper methods, therefore.

noisysocks committed:
Clear block selection when switching from template focus to page focus

noisysocks committed:
Prevent insertion into a disabled block

noisysocks committed:
Don't allow removing and moving children of disabled blocks

noisysocks committed:
Work around @wordpress/data bug by not using createRegistrySelector() for now

noisysocks commented:
> Fix selection issue in Chrome. https://github.com/WordPress/gutenberg/pull/50857#discussion_r1205058523

Stuck on this issue 😅 It happens because of a `selectionchange` event handled by `RichText`. I'm not really sure why it's not responding to `user-events: none`. I am thinking that relying on `pointer-events` and `user-events` is too fraught and my new idea is to try and design a component that places `inert` on the largest possible subtree of disabled DOM elements. I'll experiment with that in a separate PR while this one simmers.

noisysocks commented:
I opened https://github.com/WordPress/gutenberg/pull/51079 which adds `inert` to the largest possible subtree of disabled blocks.

jsnajdr commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -71,35 +71,33 @@ export function getLastInsertedBlocksClientIds( state ) {
-		}
-);
+export const getBlockEditingMode = ( state, clientId = '' ) => {
+	const explicitEditingMode = getExplcitBlockEditingMode( state, clientId );
```
You probably want to fix a typo here, adding a missing _i_ to the selector name.

ramonjd committed:
Fix typo

ramonjd commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -71,37 +71,35 @@ export function getLastInsertedBlocksClientIds( state ) {
-		}
-);
+export const getBlockEditingMode = ( state, clientId = '' ) => {
+	const explicitEditingMode = getExplicitBlockEditingMode( state, clientId );
```
The [failing unit tests](https://github.com/WordPress/gutenberg/actions/runs/5128158323/jobs/9224559312?pr=50857) in packages/block-editor/src/store/test/selectors.js are due to the state not yet being available in this private selector:

```js
	if ( ! state || ! state.blockEditingModes || ! state.blocks.parents ) {
		return 'default';
	}
```

` npm run test:unit packages/block-editor/src/store/test/selectors.js`

Does this have anything to do with private selectors not being mapped in time/no resolver support?

If so I think @jsnajdr has started looking at this: https://github.com/WordPress/gutenberg/pull/51009

noisysocks replied:
I've fixed the tests now. The tests just needed to be updated to include `blockEditingModes` in the mock state now that we're calling `getBlockEditingMode` from `canInsertBlock`.

(Highlights the frailty of writing the tests that way. We probably should update all of them to use `replaceBlocks()` to instantiate the mock state, but oh well.)

ramonjd commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1663,21 +1668,19 @@ export function canInsertBlocks( state, clientIds, rootClientId = null ) {
+	if ( getBlockEditingMode( state, rootClientId ) === 'disabled' ) {
+		return false;
+	}
+	return true;
```
What you have is easier to read, but this also works:

```js
	return (
		! attributes ||
		( ! attributes.lock?.remove &&
			! getTemplateLock( state, rootClientId ) &&
			getBlockEditingMode( state, rootClientId ) !== 'disabled' )
	);
```

I don't know why I even wrote that. To look clever I suppose.

noisysocks replied:
You look clever!

ramonjd commented on packages/block-library/src/post-title/edit.js:
```diff
@@ -114,20 +119,22 @@ export default function PostTitleEdit( {
-				/>
-			</BlockControls>
+			{ blockEditingMode === 'default' && (
+				<BlockControls group="block">
```
<img width="228" alt="Screenshot 2023-05-31 at 3 32 14 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/a840c4d1-5e17-41c1-8ec7-07a6167c9cab">

The black dot is a collapsed border style, and appears on all page blocks that are `contentOnly`, e.g., feature image block as well.

There's a permanent border on the following element (I've added `border:0` inline to remove it via the console)

```html
<div id="id-0f1fdc" role="toolbar" aria-orientation="horizontal" aria-label="Block tools" class="components-accessible-toolbar block-editor-block-contextual-toolbar" style="
    border: 0;
"><div class="block-editor-block-toolbar"><div></div></div></div>
```

This element is rendered by `<BlockContextualToolbar />` ([block-contextual-toolbar.js](https://github.com/WordPress/gutenberg/blob/9eba2068716dce9f79c1190c60ac6ff1f2de6689/packages/block-editor/src/components/block-tools/block-contextual-toolbar.js#L82-L82)). The offending CSS rule for `.block-editor-block-contextual-toolbar` is in [style.scss](https://github.com/WordPress/gutenberg/blob/9eba2068716dce9f79c1190c60ac6ff1f2de6689/packages/block-editor/src/components/block-tools/style.scss#L98-L98).

The element is always rendered, even when the slots are empty.  

I'm not sure what the best approach is. Checking via `useSlotFills()` whether the slot is, indeed filled? Would we have to loop through all [groups.js](https://github.com/WordPress/gutenberg/blob/9eba2068716dce9f79c1190c60ac6ff1f2de6689/packages/block-editor/src/components/block-controls/groups.js#L12-L12)?

I'm also guessing the value of `blockEditingMode` in  <BlockContextualToolbar />` might not be useful to since the toolbar can show for `default` and `contentOnly`. In this case the site title block has a value of `contentOnly`, but the slots are not filled.

Just dumping my brain here in case it helps... :)

noisysocks replied:
Thanks for digging into this. I'm also not sure how best to fix it. It's a shame we can't use the `:empty` pseudo selector in CSS because the toolbar technically isn't empty (there's a `<div />`). Checking if there's fills via `useSlotFills()` is probably the best bet like you say.

At any rate I'm going to leave fix this for a follow-up PR. (It's noted in the PR description.)

tellthemachines replied:
Could we do something like :has( > div:empty:only-child)? Fine to address as a follow-up though 😅

noisysocks replied:
`:has` doesn't work in Firefox 😭 I'll pick your brain about this next week.

ramonjd reviewed (commented):
Functionally, this is working pretty well. 

I know there are a few TODOs in the description to go through. Happy to help test more of them.

There's also a [suggestion](https://github.com/WordPress/gutenberg/pull/50643#pullrequestreview-1450920100) by @Mamaduka to optimize some of the selectors, one of which I [tried with some success](https://github.com/WordPress/gutenberg/pull/50643#discussion_r1211085849).

noisysocks committed:
Fix select() mock

noisysocks committed:
Fix block-editor selector tests

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into add/template-post-context-switching

noisysocks committed:
Revert block-editor changes

noisysocks commented:
I've split this large PR up a bit to make it a little less daunting. As mentioned, there's https://github.com/WordPress/gutenberg/pull/51079 which should address the issue in Chrome that folks have been noticing. I've also moved the `@wordpress/block-editor` changes into https://github.com/WordPress/gutenberg/pull/51148 since they're technically follow-up enhancements to https://github.com/WordPress/gutenberg/pull/50643. 

All three PRs ([1](https://github.com/WordPress/gutenberg/pull/51079), [2](https://github.com/WordPress/gutenberg/pull/51148), [3](https://github.com/WordPress/gutenberg/pull/50857)) can be tested independently though personally I think it's easier to `git merge` them into a single branch. Ideally we merge them all at around the same time.

noisysocks commented:
Looks like performance tests are legitimately failing. The problem is that every block calls `setBlockEditingMode()` which invalidates the cache. Do this for a couple of hundred blocks and things get real slow. I'll fix the issue in a separate PR to be merged first.

noisysocks commented:
> I'll fix the issue in a separate PR to be merged first.

Opened https://github.com/WordPress/gutenberg/pull/51185.

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into add/template-post-context-switching

noisysocks committed:
Improve useDisableNonContentBlocks performance

noisysocks committed:
Fix performance tests

tellthemachines reviewed (commented):
I just gave this a quick test and noticed it's still possible to edit the Post Content block's settings through the sidebar:

https://github.com/WordPress/gutenberg/assets/8096000/1fd8cca0-964b-479d-8490-54624e08d9dc

Could we perhaps disable the parent selector for top-level blocks inside Post Content?

noisysocks commented:
Thanks for testing @tellthemachines! Good catch. I just tested it and if I merge https://github.com/WordPress/gutenberg/pull/50912 into this branch then that fixes that issue.

tellthemachines commented on packages/edit-site/src/store/selectors.js:
```diff
@@ -321,3 +321,27 @@ export function isNavigationOpened() {
+ *
+ * @return {boolean} Whether or not the editor has a page loaded into it.
+ */
+export function isPage( state ) {
```
Should this be called `isPost` instead? It's already possible to edit a post from the site editor if it's linked to from the Navigation. Same for `hasPageContentLock` and `setHasPageContentLock` I guess 😄

noisysocks replied:
Yeah agree. Thinking I'll rename everything I'm adding in this PR to be `post` instead of `page` which better matches the WordPress convention of calling every CPT-like thing a _post_. I'll do it in a follow-up though 😅 (It's in my list in the PR's description.)

tellthemachines reviewed (commented):
This is looking really good! My only real question is about selector naming. Apart from that it's pretty much ✅

noisysocks commented:
I've merged https://github.com/WordPress/gutenberg/pull/51079, https://github.com/WordPress/gutenberg/pull/51148, https://github.com/WordPress/gutenberg/pull/51185, and https://github.com/WordPress/gutenberg/pull/50912 which are complimentary PRs to this one.

Heaps of follow-up work to do (see description) but we need contributors and users to start testing this ASAP, so let's get it in.

Thanks for the help everyone! Especially @SaxonF for all the design input and @tellthemachines and @ramonjd for all the testing.

paaljoachim commented:
I look forward to trying the feature directly in the Gutenberg plugin.

One major confusion is the editing of the Front Page template vs editing the single Front page. 
As the template consists for only one page, so it becomes unique to that one single page. Blurring the lines heavily of editing the front page template or the front page. It sure confused me if I should edit the specific page/template in the post editor or site editor.

Ren2049 commented:
[What do you think about these ideas for a visual template and post-type editor inside the site editor?](https://github.com/WordPress/gutenberg/discussions/51228) I think it could solve many problems non devs/designers have with the editor giving marketers and normal users the option to work from a higher level.
